staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -136,17 +143,32 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 					status = metrics.PostTimeoutHandlerPanic
 				}
 
-				metrics.RecordRequestPostTimeout(metrics.PostTimeoutSourceTimeoutHandler, status)
-				err := fmt.Errorf("post-timeout activity - time-elapsed: %s, %v %q result: %v",
-					time.Since(timedOutAt), r.Method, r.URL.Path, res)
+				metrics.RecordRequestPostTimeout(metricSource, status)
+				err := fmt.Errorf("post-%s activity - time-elapsed: %s, %v %q result: %v",
+					reason, time.Since(terminatedAt), r.Method, r.URL.Path, res)
 				utilruntime.HandleError(err)
 			}()
 		}()
 		httplog.SetStacktracePredicate(r.Context(), func(status int) bool {
 			return false
 		})
-		defer postTimeoutFn()
-		tw.timeout(err)
+
+		switch r.Context().Err() {
+		case context.DeadlineExceeded:
+			reason = "timeout"
+			metricSource = metrics.PostTimeoutSourceTimeoutHandler
+			responseErr = err
+			defer postTimeoutFn()
+
+		case context.Canceled:
+			reason = "cancel"
+			metricSource = metrics.PostTimeoutSourceCanceledHandler
+			responseErr = apierrors.NewBadRequest("request canceled")
```

Comment:
cc @smarterclayton for random things looking at timeout error content

Reply:
It would be better if we could return a typed error with a reason under the code.  Rational for `bad request` being the code?

Reply:
Request context being cancelled happens on the server side, right?  Or is an early close of the request actually modelled as a context cancel now in the go client library?  If it's due to an early close I can understand bad request.

If so, I'd really like to ensure (no matter what the code is) that we return a deterministic reason `metav1.StatusReasonRequestCanceled = "RequestCancelled"`.  Then I would expect our go client to potentially have a helper that can detect server cancellation and eat errors on the client side (if we requested the cancel, and we get a cancel, eat that error, otherwise someone else cancelled, show the error).

Reply:
I'm pretty sure this early close initiated by the client side (you can see how the unit test implements it), but I'll try and confirm this. Given it's closed client-side, I don't think the error we return here is ever going to reach the client, so it's primary purpose is server-side logging & monitoring.

Reply:
Right, but I think this will turn up in audit log (maybe?), so want a deterministic reason.  And for other contexts like exec sessions, log streaming, watches, websocket watches, and maybe even proxies I could still see there being a need for this where client initiated cancellation doesn't implicitly close the response channel.

Reply:
Sorry, just getting back to this (was on leave). At least in the Go client, I think timeout & client side cancel will both result in `client.Do` returning immediately and ignoring the server-side response, so I don't think changing the response error code here will affect client behaviour.

The value to this change is mostly for server side metrics ("are too many requests timing out?"), and audit & debug logs. That said, the race condition raised in https://github.com/kubernetes/kubernetes/pull/116871#issuecomment-1505553888 still stands. I'm not sure that there's a good way to resolve that, but treating timeouts as a client-side cancel seems more correct than treating a client-side cancel as a timeout.